### PR TITLE
add create-route form component

### PR DESCRIPTION
**Pull request by divyanshiGupta**
Story: https://jira.coreos.com/browse/ODC-818

This PR adds a reusable create-route form that can be integrated across all create flows. Currently only path and host-name are being validated as a part of basic validations and later on the entire `ImportFromGit` form will be validated by doing a dry run on everything first and if there are any errors present, those errors will be shown. 
For reference see https://coreos.slack.com/archives/C6A3NV5J9/p1559577335051200

### CODE REVIEW

File: frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+          'Hostname must consist of lower-case letters, numbers, periods, and hyphens. It must start and end with a letter or number.';
+      } else if (hostname.length > hostNameValidationParams.maxLength) {
+        hostnameErrorMsg = `Can't be longer than ${hostNameValidationParams.maxLength} characters.`;
+      } else {
+        hostnameErrorMsg = '';
+      }
+    }
+  };
+
+  const isPathValid = () => {
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx]: Move to a separate utility class so testing is easier.

File: frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+  const hostNameValidationParams = {
+    pattern: /^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$/,
+    maxLength: 253,
+  };
+
+  const pathValidationParams = {
+    pattern: /^\/.*$/,
+  };
+
+  const isHostNameValid = () => {
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx]: Move to a separate utility class so testing is easier. Also the parent should own the validation so that it can control the state of the form.

File: frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+                  <HelpBlock>{hostnameErrorMsg}</HelpBlock>
+                </div>
+              )}
+            </FormGroup>
+            <FormGroup>
+              <ControlLabel>Path</ControlLabel>
+              <FormControl
+                type="text"
+                onChange={handleChange}
+                onBlur={isPathValid()}
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx]: This isn't right because this function returns `undefined`.
You do not need an `onBlur`

File: frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+  path: string;
+  loadedServices?: boolean;
+  serviceOptions?: {};
+  portOptions: {};
+  service?: K8sResourceKind;
+  targetPort: string;
+  createRoute?: boolean;
+  createRouteOption?: boolean;
+  helpText?: boolean;
+  errorText?: boolean;
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx]: When would `helpText` or `errorText` ever be set to `false`?

File: frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+  handleChange: (event: React.SyntheticEvent<HTMLInputElement>) => void;
+  changeService?: (selectedKey: string) => void;
+  changeTargetPort: (selectedKey: string) => void;
+  toggleSection?: (event: React.SyntheticEvent<HTMLInputElement>) => void;
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx]: Callbacks should start with `on`. eg `onChange`

Callbacks should be required.

File: frontend/public/components/routes/create-route.tsx [hunk 6]
```diff
@@ -229,141 +229,46 @@ export class CreateRoute extends React.Component<null, CreateRouteState> {
...
-              {termination === 'reencrypt' && <div className="form-group co-create-route__destinationCaCertificate">
-                <DroppableFileInput
-                  onChange={this.onDestinationCACertificateChange}
-                  inputFileData={this.state.destinationCACertificate}
-                  id="destination-ca-certificate"
-                  label="Destination CA Certificate" />
-              </div>}
-            </React.Fragment>}
-          </div> }
+        <Form onSubmit={this.save} className="co-m-pane__body-group">
```
Comment (christianvogt) [inline frontend/public/components/routes/create-route.tsx]: Why `Form` and why the change in className?

File: frontend/public/components/routes/create-route.tsx [hunk 7]
```diff
@@ -229,141 +229,46 @@ export class CreateRoute extends React.Component<null, CreateRouteState> {
...
+            loadedServices={loaded}
+            portOptions={portOptions}
+            handleChange={this.handleChange}
+            changeService={this.changeService}
+            changeTargetPort={this.changeTargetPort}
+          />
+          <SecureRouteForm
+            termination={termination}
+            certificate={certificate}
+            key={key}
```
Comment (christianvogt) [inline frontend/public/components/routes/create-route.tsx]: `key` is a bad property too use as it's the name of a standard react prop. This should be renamed.

File: frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx [hunk 8]
```diff
@@ -0,0 +1,205 @@
+import * as React from 'react';
+import * as _ from 'lodash-es';
+import { FormControl, FormGroup, ControlLabel, HelpBlock } from 'patternfly-react';
+import { Dropdown } from '../../../../components/utils';
+import { K8sResourceKind } from '../../../../module/k8s';
+
+interface CreateRouteFormProps {
+  name?: string;
+  hostname: string;
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx]: I think every field should be optional.

File: frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+            {name !== undefined && (
+              <FormGroup>
+                <ControlLabel className="co-required">Name</ControlLabel>
+                <FormControl
+                  type="text"
+                  onChange={handleChange}
+                  value={name}
+                  placeholder="my-route"
+                  id="name"
+                  name="name"
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx:74]: Id is too generic. Ideally we should generate these.
The name is also quite generic and can clash with other form fields.
@rohitkrai03 will your form components generate IDs and scope names?

File: frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+                  aria-describedby="name-help"
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx:75]: Where is `name-help`?

File: frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+                  onChange={handleChange}
+                  onBlur={isHostNameValid()}
+                  value={hostname}
+                  placeholder="www.example.com"
+                  id="hostname"
+                  name="hostname"
+                  aria-describedby="hostname-help"
+                />
+              {helpText && (
+                <HelpBlock>
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx]: I think `HelpBlock`s would need an ID for the `aria-describedby` attribute to properly reference the help.

File: frontend/public/components/routes/create-route.tsx (hunk 6 above)
Comment (divyanshiGupta) [reply frontend/public/components/routes/create-route.tsx]: Because we are using `Form` from pf3 and also when we will migrate to pf4 we will just have to change the import. I changed the className because we are using "co-m-pane_body-group" in our forms so just for consistency i changed it.

File: frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx (hunk 9 above)
Comment (rohitkrai03) [reply frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx:74]: @christianvogt I've added a way to add unique ids to form fields but that is only unique to the form you are in and not the whole app. For that maybe i need to generate uuids.

File: frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx (hunk 1 above)
Comment (divyanshiGupta) [reply frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx]: Yes, so it was discussed that we will be having a separate class for all the validations but for now I added it here because all of the other forms are going to have validations and we should not create separate classes for each form. I thought once we are working integrating all the parts together we can do this as well at that time

File: frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx (hunk 10 above)
Comment (divyanshiGupta) [reply frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx]: I didnt add it because we dont need name in our form and in console form help text is not displayed

**Comment by rohitkrai03** [comment]
build and deploy

**Comment by invincibleJai** [comment]
build and deploy

**Comment by sbose78** [comment]
Your build is progress : https://console-openshift-console.apps.rohit31.devcluster.openshift.com/k8s/ns/openshift-console/builds
 
 PR will be LIVE in https://pr-122-openshift-console.apps.rohit31.devcluster.openshift.com

File: frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx (hunk 5 above)
Comment (divyanshiGupta) [reply frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx]: Yes, but in this case we need  `toggleSection` only when we have `create-route` option inside the form as a checkbox. Form in console doesn't require this.

File: frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx (hunk 10 above)
Comment (divyanshiGupta) [reply frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx:75]: Added

File: frontend/public/components/routes/create-route.tsx (hunk 6 above)
Comment (christianvogt) [reply frontend/public/components/routes/create-route.tsx]: `co-create-route` has rules associated with it. Are they no longer required? I'd be worried to just remove this.
Furthermore `co-m-pane_body-group` should be `co-m-pane__body-group`. Why do we need this?
Also `deploy-image.tsx` uses `co-m-pane__form` just FYI

File: frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+  portOptions?: {};
+  service?: K8sResourceKind;
+  targetPort?: string;
+  createRoute?: boolean;
+  createRouteOption?: boolean;
+  errorText?: boolean;
+  onInputChange?: (event: React.SyntheticEvent<HTMLInputElement>) => void;
+  onServiceChange?: (selectedKey: string) => void;
+  onTargetPortChange?: (selectedKey: string) => void;
+  onSectionToggle?: (event: React.SyntheticEvent<HTMLInputElement>) => void;
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx]: `onSectionToggle` should be named according to the value that drives it.
Use `onCreateRouteChange`.

Should probably have a boolean param and not the event itself.

File: frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+                  Public hostname for the route. If not specified, a hostname is generated.
+                </HelpBlock>
+              </FormGroup>
+            )}
+            <FormGroup>
+              <ControlLabel>Hostname</ControlLabel>
+              <FormControl
+                type="text"
+                onChange={onInputChange}
+                onBlur={isHostNameValid()}
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx]: Remove this.

**Comment by christianvogt** [comment]
Can you squash the changes to the console create route file into a separate commit please. It'll make merging to console repo easier.

File: frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx (hunk 12 above)
Comment (divyanshiGupta) [reply frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx]: So this event is called when the user clicks on the `checkbox` and it sets `checked` as true in parent's `state`. `checked` is a boolean prop that is used to `check`/`uncheck` the `checkbox`.

**Comment by divyanshiGupta** [comment]
> Can you squash the changes to the console create route file into a separate commit please. It'll make merging to console repo easier.

Done.

**Comment by divyanshiGupta** [comment]
@christianvogt I have removed the changes from `create-route` of console because once we change the `CreateRouteForm` to use `Formik` fields as part of the integration story it will break in `create-route`. Integrating `CreateRouteForm` in `create-route` and converting `create-route` component to a `Formik` form should be a part of the refactoring story.

File: frontend/public/extend/devconsole/shared/utils/create-route-utils.ts [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+  pattern:
+    'Hostname must consist of lower-case letters, numbers, periods, and hyphens. It must start and end with a letter or number.',
+  maxLength: `Can't be longer than ${hostNameValidationParams.maxLength} characters.`,
+};
+
+export const pathErrorMessages = {
+  pattern: 'Path must start with /',
+};
+
+export const validateHostName = (hostname: string): string => {
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/utils/create-route-utils.ts:21]: I think these should return `string?` so that we can return `undefined` rather than an empty string.
Not sure what best aligns with the overall validation strategy so I'm fine to leave this for now

File: frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+import { Dropdown } from '../../../../components/utils';
+import { K8sResourceKind } from '../../../../module/k8s';
+
+interface CreateRouteFormProps {
+  name?: string;
+  hostname?: string;
+  path?: string;
+  loadedServices?: boolean;
+  serviceOptions?: {};
+  portOptions?: {};
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx]: Options should be properly typed.

File: frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+  service?: K8sResourceKind;
+  targetPort?: string;
+  createRoute?: boolean;
+  createRouteOption?: boolean;
+  hostnameErrorMsg?: string;
+  pathErrorMsg?: string;
+  onInputChange?: (event: React.SyntheticEvent<HTMLInputElement>) => void;
+  onServiceChange?: (selectedKey: string) => void;
+  onTargetPortChange?: (selectedKey: string) => void;
+  onCreateRouteChange?: (event: React.SyntheticEvent<HTMLInputElement>) => void;
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx:23]: Param value is fine for now.
@rohitkrai03 please align all callbacks with appropriate types as we integrate forms. Should they provide event? or value directly?

Comment (rohitkrai03) [reply frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx:23]: @christianvogt Most of the event handlers for HTML input elements would be automatically handled by the wrapper component. The custom components will need to have its own set of fields and values as per the need. Most commonly custom component will be using `setFieldValue` to set state.

File: frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx (hunk 15 above)
Comment (divyanshiGupta) [reply frontend/public/extend/devconsole/shared/components/CreateRouteForm.tsx]: Done

**Review by rohitkrai03 — APPROVED** [review]
LGTM
